File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/SubarraySum.java
package com.only4.algorithm.version4j.leetcode;

import java.util.HashMap;
import java.util.Map;

public class SubarraySum {
    public int subarraySum(int[] nums, int k) {
        // 记录前缀和的出现次数，初始化前缀和0出现1次（空数组）
        Map<Integer, Integer> prefixSumCount = new HashMap<>();
        prefixSumCount.put(0, 1);

        int count = 0;
        int currentPrefixSum = 0;

        for (int num : nums) {
            // 计算当前前缀和
            currentPrefixSum += num;

            // 查找是否存在前缀和为 currentPrefixSum - k
            // 如果存在，说明有子数组的和为k
            count += prefixSumCount.getOrDefault(currentPrefixSum - k, 0);

            // 更新当前前缀和的出现次数
            prefixSumCount.put(currentPrefixSum,
                    prefixSumCount.getOrDefault(currentPrefixSum, 0) + 1);
        }

        return count;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/subarraySum.kt
package com.only4.algorithm.version4k.leetcode

fun subarraySum(nums: IntArray, k: Int): Int {
    // 记录前缀和的出现次数，初始化前缀和0出现1次（空数组）
    val prefixSumCount = mutableMapOf<Int, Int>().withDefault { 0 }
    prefixSumCount[0] = 1

    var count = 0
    var currentPrefixSum = 0

    for (num in nums) {
        // 计算当前前缀和
        currentPrefixSum += num

        // 查找是否存在前缀和为 currentPrefixSum - k
        // 如果存在，说明有子数组的和为k
        count += prefixSumCount.getValue(currentPrefixSum - k)

        // 更新当前前缀和的出现次数
        prefixSumCount[currentPrefixSum] = prefixSumCount.getValue(currentPrefixSum) + 1
    }

    return count
}
